Examine non-deterministic test-failure in release/1.0 branch

E.g. this one: https://travis-ci.org/spray/spray/jobs/10496447

This could be a real race-condition that for some reason is exposed by that specific configuration. For me it seems like `HttpHostConnectionSlot` isn't correctly reacting to it's underlying connection dying (or forgets to unwatch the connection from a previous try?) in at least of its states leading to the slot dying with a `DeathPactException`. The next retry then seems to send out the `Connect` correctly but the `Connected` message never reaches its target because the slot then already has died.

```
[WARN] [08/24/2013 14:08:59.724] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-22] [akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/host-connector-6/0] GET request to /asdsadsadsasdaasd timed out in response to GET request to /asdsadsadsasdaasd with 1 retries left, retrying...
[DEBUG] [08/24/2013 14:08:59.724] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-22] [akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/host-connector-6/0] Attempting new connection to localhost:57289
[DEBUG] [08/24/2013 14:08:59.725] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-24] [akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/group-0/14] Attempting connection to localhost/127.0.0.1:57289
[DEBUG] [08/24/2013 14:08:59.725] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-21] [akka://spray-can-client-SprayCanClientSpec/system/IO-TCP/selectors/$a/42] Attempting connection to [localhost/127.0.0.1:57289]
[ERROR] [08/24/2013 14:08:59.725] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-22] [akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/host-connector-6/0] Monitored actor [Actor[akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/group-0/6]] terminated
akka.actor.DeathPactException:Monitored actor [Actor[akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/group-0/6]] terminated
[f56afc10-0cb5-11e3-b9f3-84a6c82bf84f]

[DEBUG] [08/24/2013 14:08:59.729] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-21] [akka://spray-can-client-SprayCanClientSpec/system/IO-TCP/selectors/$a/42] Connection established to [localhost/127.0.0.1:57289]
[DEBUG] [08/24/2013 14:08:59.729] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-24] [akka://spray-can-client-SprayCanClientSpec/system/IO-TCP/selectors/$a/12] New connection accepted
[DEBUG] [08/24/2013 14:08:59.729] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-26] [akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/group-0/14] Connected to localhost/127.0.0.1:57289
[WARN] [08/24/2013 14:08:59.729] [spray-can-client-SprayCanClientSpec-akka.actor.default-dispatcher-17] [akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/host-connector-6/0] received dead letter from Actor[akka://spray-can-client-SprayCanClientSpec/user/IO-HTTP/group-0/14]: Connected(localhost/127.0.0.1:57289,/127.0.0.1:36475)
```

--- a/spray-can/src/main/scala/spray/can/client/HttpHostConnectionSlot.scala
+++ b/spray-can/src/main/scala/spray/can/client/HttpHostConnectionSlot.scala
@@ -57,10 +57,6 @@ private[client] class HttpHostConnectionSlot(host: String, port: Int,
           case ctx: RequestContext  ⇒ context.parent ! ctx
           case _: Http.CloseCommand ⇒ context.stop(self)
         }
-
-      case Terminated(conn) ⇒
-      // ignore, may happen if in closing we unwatch too late
-      // and this message is already in the mailbox
     }
   }
 
@@ -134,18 +130,16 @@ private[client] class HttpHostConnectionSlot(host: String, port: Int,
       context.become(terminating(httpConnection))
 
     case ev: Http.ConnectionClosed ⇒
-      context.parent ! Disconnected(openRequests.size)
+
       val errorMsgForOpenRequests = ev match {
         case Http.PeerClosed ⇒ "Premature connection close (the server doesn't appear to support request pipelining)"
         case x               ⇒ x.toString
       }
-      openRequests foreach clear(errorMsgForOpenRequests, retry = true)
-      context.unwatch(httpConnection)
-      context.become(unconnected)
+      reportDisconnection(openRequests, errorMsgForOpenRequests, retry = true)
+      context.become(waitingForConnectionTermination(httpConnection))
 
     case Terminated(`httpConnection`) ⇒
-      context.parent ! Disconnected(openRequests.size)
-      openRequests foreach clear("Unexpected connection termination", retry = true)
+      reportDisconnection(openRequests, "Unexpected connection termination", retry = true)
       context.become(unconnected)
   }
 
@@ -154,21 +148,37 @@ private[client] class HttpHostConnectionSlot(host: String, port: Int,
 
   def closing(httpConnection: ActorRef, openRequests: Queue[RequestContext], error: Http.ConnectionException,
               retry: Boolean): Receive = {
+    case _: Http.ConnectionClosed ⇒
+      reportDisconnection(openRequests, error, retry)
+      context.become(waitingForConnectionTermination(httpConnection))
 
-    case ev @ (_: Http.ConnectionClosed | Terminated(`httpConnection`)) ⇒
-      context.parent ! Disconnected(openRequests.size)
-      openRequests foreach clear(error, retry)
-      context.unwatch(httpConnection)
+    case Terminated(`httpConnection`) ⇒
+      reportDisconnection(openRequests, error, retry)
       context.become(unconnected)
+  }
+  def waitingForConnectionTermination(httpConnection: ActorRef): Receive = {
+    var buffered = Queue.empty[Any]
+
+    {
+      case Terminated(`httpConnection`) ⇒
+        context.become(unconnected)
+        buffered.foreach(self ! _)
+      case x ⇒ buffered = buffered.enqueue(x)
+    }
   }
 
   def terminating(httpConnection: ActorRef): Receive = {
     case _: Http.ConnectionClosed     ⇒ // ignore
     case Terminated(`httpConnection`) ⇒ context.stop(self)
   }
+  def reportDisconnection(openRequests: Queue[RequestContext], error: String, retry: Boolean): Unit =
+    reportDisconnection(openRequests, new Http.ConnectionException(error), retry)
+  def reportDisconnection(openRequests: Queue[RequestContext], error: Http.ConnectionException, retry: Boolean): Unit = {
+    context.parent ! Disconnected(openRequests.size)
+    openRequests foreach clear(error, retry)
+  }
 
   def clear(error: String, retry: Boolean): RequestContext ⇒ Unit = clear(new Http.ConnectionException(error), retry)
-
   def clear(error: Http.ConnectionException, retry: Boolean): RequestContext ⇒ Unit = {
     case ctx @ RequestContext(request, retriesLeft, _) if retry && request.canBeRetried && retriesLeft > 0 ⇒
       log.warning("{} in response to {} with {} retries left, retrying...", error.getMessage, format(request), retriesLeft)

